sw/device/tests/sim_dv/i2c_device_tx_rx_test.c

```diff
@@ -0,0 +1,225 @@
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sw/device/lib/testing/autogen/isr_testutils.h"
+
+// TODO, remove it once pinout configuration is provided
+#include "pinmux_regs.h"
+
+static dif_i2c_t i2c;
+static dif_pinmux_t pinmux;
+static dif_rv_plic_t plic;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+/**
+ * This symbol is meant to be backdoor loaded by the testbench.
+ * The testbench will inform the test the rough speed of the clock
+ * used by the I2C modules.
+ *
+ * The I2C Device state machine does depend on the I2C timing configuration
+ */
+static volatile const uint8_t kClockPeriodNanos = 0;
+static volatile const uint8_t kI2cRiseFallNanos = 0;
+static volatile const uint32_t kI2cClockPeriodNanos = 0;
+
+/**
+ * This symbol is meant to be backdoor loaded by the testbench.
+ * to indicate which I2c is actually under test. It is not used
+ * at the moment, will connect it later.
+ */
+static volatile const uint8_t kI2cIdx = 0;
+
+/**
+ * This set of symbols is meant to be backdoor loaded by the testbench.
+ * to indicate the address that will be listened to by the device.
+ */
+static volatile const uint8_t kI2cDeviceAddress0 = 0x55;
+static volatile const uint8_t kI2cDeviceMask0 = 0x7f;
+static volatile const uint8_t kI2cDeviceAddress1 = 0x7f;  // disable match on
+                                                          // second address
+static volatile const uint8_t kI2cDeviceMask1 = 0x00;
+
+/**
+ * This symbol is meant to be backdoor loaded by the testbench.
+ * to indicate the number of bytes that should be sent.
+ *
+ * Because the test doesn't manage the FIFO during writes, it's unprepared to
+ * recieve more than the address + 62 bytes + junk
+ */
+static volatile const uint8_t kI2cByteCount = 0;
+
+static volatile bool tx_empty_irq_seen = false;
+static volatile bool trans_complete_irq_seen = false;
+
+/**
+ * Provides external irq handling for this test.
+ *
+ * This function overrides the default OTTF external ISR.
+ */
+void ottf_external_isr(void) {
+  plic_isr_ctx_t plic_ctx = {.rv_plic = &plic,
+                             .hart_id = kTopEarlgreyPlicTargetIbex0};
+
+  i2c_isr_ctx_t i2c_ctx = {
+      .i2c = &i2c,
+      .plic_i2c_start_irq_id = kTopEarlgreyPlicIrqIdI2c0FmtWatermark,
+      .expected_irq = 0,
+      .is_only_irq = false};
+
+  top_earlgrey_plic_peripheral_t peripheral;
+  dif_i2c_irq_t i2c_irq;
+  isr_testutils_i2c_isr(plic_ctx, i2c_ctx, &peripheral, &i2c_irq);
+
+  switch (i2c_irq) {
+    case kDifI2cIrqTxEmpty:
+      tx_empty_irq_seen = true;
+      i2c_irq = kDifI2cIrqTxEmpty;
+      break;
+    case kDifI2cIrqTransComplete:
+      trans_complete_irq_seen = true;
+      i2c_irq = kDifI2cIrqTransComplete;
+      break;
+    default:
+      LOG_ERROR("Unexpected interrupt (at I2C): %d", i2c_irq);
+      break;
+  }
+}
+
+bool test_main(void) {
+  if (kI2cByteCount > 62) {
```

Comment:
Yeah, I can do min(I2C_TX_FIFO_DEPTH, I2C_ACQ_FIFO_DEPTH) - 2

Reply:
Looks like the software is only passed I2C_PARAM_FIFO_DEPTH and like I did my math badly here because we have the 2 transactions in the test, but it's now fixed in the code.